test/tests/functional/pbs_job_array.py

```diff
@@ -68,3 +69,51 @@ def test_arrayjob_Erecord_startval(self):
         # Verify that the value of 'start' isn't 0
         self.assertNotEqual(start_val, 0,
                             "E record value of 'start' for arrayjob is 0")
+
+    def test_arraysubjob_survive_restart(self):
+        """
+        Test to check if subjobs of an array job survive a pbs_server restart
+        """
+        a = {'resources_available.ncpus': 1}
+        self.server.manager(MGR_CMD_SET, NODE, a, self.mom.shortname)
+        j = Job(TEST_USER, attrs={
+            ATTR_J: '1-3', 'Resource_List.select': 'ncpus=1'})
+
+        j.set_sleep_time(5)
+
+        j_id = self.server.submit(j)
+
+        # 1. check job array has begun
+        self.server.expect(JOB, {'job_state': 'B'}, j_id)
+
+        # 2. wait till subjob 2 starts running
+        self.server.expect(JOB, {'job_state': 'R'}, create_subjob_id(j_id, 2))
+
+        # 3. Kill the server
+        self.server.stop('-KILL')
+
+        # 4. start the server
+        self.server.start()
+
+        # 5. array job should be B
+        self.server.expect(JOB, {'job_state': 'B'}, j_id, max_attempts=1)
+
+        # 6. subjob 1 should be X
+        self.server.expect(JOB, {'job_state': 'X'}, create_subjob_id(j_id, 1),
+                           max_attempts=1)
+
+        # 7. subjob 2 should be R
+        self.server.expect(JOB, {'job_state': 'R'}, create_subjob_id(j_id, 2),
+                           max_attempts=1)
+
+        # 8. subjob 3 should be Q
+        self.server.expect(JOB, {'job_state': 'Q'}, create_subjob_id(j_id, 3),
+                           max_attempts=1)
+
+
+def create_subjob_id(job_array_id, subjob_index):
```

Comment:
I see a copy of this function already present in the test suite TestJobArrayComment. We are defining it here. I suggest you to move this function to pbs_testlib.py.

Reply:
fixed in rc 22 https://github.com/PBSPro/pbspro/pull/590/commits/5fc6c5f74588a21c53052858c0900da053784f7b